JBPM-7601 (Stunner) - Case Files

Josephblt opened the pull request:
https://issues.jboss.org/browse/JBPM-7601

This adds case file variables to the new Designer. They idea is to
replicate the same functionality of the old Designer. Other than minimal
UI differences they both will work identically.

The variables will be displayed in a different section called Case Data,
under the Process Data section, where available. Case File Variables can
be used on the MI Collection Input/Output in Multiple Instance Processes
as well.

Josephblt commented:
Jenkins execute full downstream build

Josephblt commented:
Hey @krisv, @ibek, @tiagodolphine.
Hey guys, if you could please take a look at this, that would be great. Thanks.

hasys commented:
Hi @Josephblt,

thank you, looks good. But I found two issues:
1. For old marshaller Case File Variables are not persist. If you save/reopen process you will loose all defined Case File Variables.
2. It is not possible to use Process/Sub-process Case File variables in tasks as Source or Target. It is not described in Jira issue, but I compared old and new Designers and it looks like it should be supported. See pictures below (for old and for new Designer):

Data Assignments for User Task in **Old** Designer:
![task_has_case_file_variable](https://user-images.githubusercontent.com/1477262/45216545-46141480-b2a1-11e8-976e-f0daf1bfa061.png)

Data Assignments for User Task in **New** Designer:
![screenshot from 2018-09-07 12-30-03](https://user-images.githubusercontent.com/1477262/45216626-8ffcfa80-b2a1-11e8-90f2-d34c5e279b62.png)


P.S. @krisv @ibek Case File variables are visible for both project type: BPMN and Case in old designer as well as for new designer. So, I am little bit confused here and it is not clear to me:

- Is _Process Variables_ section should be displayed in Case Project only?
- Is _Case File Variables_ section should be displayed in Default Project only?
- What about variables inside of task (for both Process Variables and Case File variables) 

P.P.S @Josephblt going to check the code.

krisv commented:
1. I don't think we want to go back to the old parser and extend that, it's only there still in case there are severe issues with the new parser, but it should be removed soon.  
2. Case file variables should indeed also be listed as possible sources and targets in the parameter mapping of nodes (human tasks, service tasks, reusable sub-process, etc.).
3. Case roles and case file variables should only be visible for case projects, process variables should be visible for both.  But afaik we are still working on this mechanism.

hasys commented:
Hi @krisv thank you for the details!

@Josephblt any news here?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/AdHocSubprocess.java:
```diff
@@ -69,14 +70,23 @@
+            afterElement = "processData"
+    )
+    @Valid
+    private CaseData caseData;
```
@Josephblt I think this `CaseData` class is the same one as @tiagodolphine already introduced as for the specific case management's property set, isn't it? I mean, are you refactoring this? It's also been added as a field in all bean implementations for subprocesses, can we move it to the abstraction class?

Josephblt commented:
Jenkins execute full downstream build

Josephblt commented:
Hey @tiagodolphine, @hasys, @romartin. Rebased everything and removed some of it, just like we talked today. Case Files are available only for the diagram.

tiagodolphine commented:
Jenkins execute full downstream build

hasys commented:
Hi @Josephblt,

looks like issues I mentioned above still aren't fixed see https://github.com/kiegroup/kie-wb-common/pull/2089#issuecomment-419411923. Is it discussed  part as you mentioned on your last comment?

Josephblt commented:
Hey @hasys. Yes, I just checked it and it is missing. I'm fixing it right now.

Josephblt commented:
Jenkins execute full downstream build

Josephblt commented:
Hey @krisv, @ibek, @romartin.
Hey guys, if you could please take a look at this, that would be great. Thanks.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/cm/CaseFileVariables.java:
```diff
@@ -0,0 +1,91 @@
+import org.kie.workbench.common.stunner.core.definition.annotation.property.Type;
+import org.kie.workbench.common.stunner.core.definition.annotation.property.Value;
+import org.kie.workbench.common.stunner.core.definition.property.PropertyType;
+
```
```suggestion
import org.kie.workbench.common.stunner.core.util.HashUtil;

```

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/cm/CaseManagementSet.java:
```diff
@@ -62,16 +72,26 @@ public void setCaseRoles(CaseRoles caseRoles) {
         if (o instanceof CaseManagementSet) {
             CaseManagementSet other = (CaseManagementSet) o;
-            return caseRoles.equals(other.caseRoles);
+            return caseRoles.equals(other.caseRoles) &&
```
```suggestion
            return Objects.equals(caseRoles, other.caseRoles)  &&
                    Objects.equals(caseFileVariables, other.caseFileVariables);
```

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/processes/RootProcessConverter.java:
```diff
@@ -76,6 +77,9 @@ private ProcessPropertyWriter convertProcessNode(Node<Definition<BPMNDiagramImpl
         p.setCaseRoles(caseRoles);
 
+        final CaseFileVariables caseFileVariables = definition.getCaseManagementSet().getCaseFileVariables();
+        p.setCaseFileVariables(caseFileVariables);
```
There is no test for this change.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/properties/ProcessPropertyWriter.java:
```diff
@@ -187,6 +188,21 @@ public void setProcessVariables(ProcessVariables processVariables) {
         });
     }
 
+    public void setCaseFileVariables(CaseFileVariables caseFileVariables) {
```
This method is not tested.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/tostunner/properties/CaseFileVariableReader.java:
```diff
@@ -0,0 +1,57 @@
+import org.eclipse.bpmn2.Property;
+import org.kie.workbench.common.stunner.bpmn.definition.property.cm.CaseFileVariables;
+
+class CaseFileVariableReader {
```
This new class has no tests.

tiagodolphine replied:
I agree I think it would be good to add tests for this class.

romartin commented:
hey @Josephblt @hasys @tiagodolphine , how's this? :) This is starting to be urgent... 
IIRC once this merged, we'll be done for all tasks about the CM integration, right? 
Thanks!

tiagodolphine commented:
Jenkins execute full downstream build

tiagodolphine commented:
the war is not available on jenkins anymore, I'm going to trigger the full downstream again.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/tostunner/properties/CaseFileVariableReader.java:
```diff
@@ -0,0 +1,57 @@
+                .orElse(caseFileVariableName);
+    }
+
+    public static String getCaseFileVariableName(Property p) {
```
this could be private, or is it used on other classes?

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/tostunner/properties/CaseFileVariableReader.java:
```diff
@@ -0,0 +1,57 @@
+
+    public static boolean isCaseFileVariable(Property p) {
+        String name = getCaseFileVariableName(p);
+        return name.startsWith(CaseFileVariables.CASE_FILE_PREFIX);
```
any chance the `name` is returned as null, in this case, a null check before `startsWith` is great to avoid a NPE.

Josephblt replied:
Hey @tiagodolphine. The name won't be null in this case as it either has an id or a name, so no need for the extra check.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/VariablesProvider.java:
```diff
@@ -68,16 +69,34 @@ public String getProviderName() {
         return result;
     }
+
+    private static void addPropertyVariableToResult(Collection<Pair<Object, String>> result, String element) {
```
I think this method has no need to be `static`

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/VariablesProvider.java:
```diff
@@ -68,16 +69,34 @@ public String getProviderName() {
+        }
+    }
+
+    private static void addCaseFileVariableToResult(Collection<Pair<Object, String>> result, String element) {
```
same, I think this method has no need to be `static`

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/assignmentsEditor/AssignmentsEditorWidget.java:
```diff
@@ -322,6 +324,16 @@ protected String getProcessVariables() {
                         variables.append(processVariables.getValue());
                     }
+                    CaseManagementSet caseManagementSet = bpmnDiagram.getCaseManagementSet();
+                    if (caseManagementSet != null) {
```
just minor, but this if all the conditions are added on just one if sentence I think it is more clear, like:
```
if (bpmnDiagram.getCaseManagementSet() !=null 
 &&  bpmnDiagram.getCaseManagementSet().getCaseFileVariables() !=null) 
 && variables.length() > 0) {
...}
```

Josephblt replied:
In this case I'm actually using the assigned variables. So to avoid making redundant calls I'll leave it the way it is.

Josephblt commented:
Jenkins execute full downstream build

Josephblt commented:
Hey @hasys, @tiagodolphine, @romartin. I think that's it. Can you guys review it. Thanks.

romartin commented:
it looks good to me (again), thanks @Josephblt 
Waiting for @hasys and @tiagodolphine 's reviews

hasys commented:
Hi @Josephblt, looks good now, just some clean up and addition minor test coverage you can found here: https://github.com/Josephblt/kie-wb-common/pull/1

Also, there is one failing test now, I guess it is caused by @tiagodolphine merged PR about Case Roles and there are some conflicts now or something. 
<details>
  <summary>
org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverterTest.convertProcessWithCaseRoles</summary>
  <pre>Error Message
Argument(s) are different! Wanted:
processPropertyWriter.setCaseFileVariables(
    caseFileVariables
);
-> at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverterTest.convertProcessWithCaseRoles(RootProcessConverterTest.java:110)
Actual invocation has different arguments:
processPropertyWriter.setCaseFileVariables(
    null
);
-> at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverter.convertProcessNode(RootProcessConverter.java:81)
Stacktrace
org.mockito.exceptions.verification.junit.ArgumentsAreDifferent: 
Argument(s) are different! Wanted:
processPropertyWriter.setCaseFileVariables(
    caseFileVariables
);
-> at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverterTest.convertProcessWithCaseRoles(RootProcessConverterTest.java:110)
Actual invocation has different arguments:
processPropertyWriter.setCaseFileVariables(
    null
);
-> at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverter.convertProcessNode(RootProcessConverter.java:81)

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverterTest.convertProcessWithCaseRoles(RootProcessConverterTest.java:110)</pre>
</details>

Josephblt commented:
Jenkins execute full downstream build

Josephblt commented:
@hasys. Fixed the test and merged your changes. Thanks

romartin commented:
Please @Josephblt remember to fix those conflicts so @LuboTerifaj or @hasys can do the review for this. This is a blocker for 7.2 release. Thanks!

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins retest please

Josephblt commented:
Jenkins execute full downstream build

Josephblt commented:
@hasys, @tiagodolphine, @romartin. All set. I think this is ready to merge. @hasys the set method changes  you proposed on CaseFileVariables and CaseManagementSet we're left out because they caused problems related to Errai all others were merged.
